Implement egress gateway datapath

anfernee opened the pull request:
On bpf_lxc side, it uses source and destination IP to lookup the egress map. If it matches any entry in the egress map, depending on tunnel endpoint, it decides whether to foward the traffic to a remote gateway, or directly redirect to its external interface if it's gateway itself.
    
On bpf_host side, before the packet leaves the cluster, SNAT the packet with egress ip defined in the egress map.
    
The feature is guarded by ENABLE_EGRESS_GATEWAY.
    
Towards: #13575

anfernee committed:
Implement egress gateway datapath

On bpf_lxc side, it uses source and destination IP to lookup the egress
map. If it matches any entry in the egress map, depending on tunnel
endpoint, it decides whether to foward the traffic to a remote gateway,
or directly redirect to its external interface if it's gateway itself.

On bpf_host side, before the packet leaves the cluster, SNAT the packet
with egress ip defined in the egress map.

The feature is guarded by ENABLE_EGRESS_GATEWAY.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -679,6 +679,34 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 		}
 	}
 
+#ifdef ENABLE_EGRESS_GATEWAY
```
@anfernee Could you also update `bpf/Makefile` to add this macro to `LXC_OPTIONS` and `MAX_LXC_OPTIONS`? That will enable compilation and complexity tests.

anfernee committed:
Turn on egress datapath when enable-egress-gateway

Connect --enable-egress-gateway with ebpf macro ENABLE_EGRESS_GATEWAY.
The patch also fixes an issue in init.sh when TUNNEL_MODE isn't <nil>,
it should also setup tunnel endpoint.

--enable-egress-gateway requires --enable-bpf-masquerade=true because it
uses ebpf on external interface to masquerade with egress IP address.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -679,6 +679,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
+		 * direct to external interface.
+		 */
+		ret = encap_and_redirect_lxc(ctx, info->tunnel_endpoint, encrypt_key,
+					     &key, SECLABEL, monitor);
```
This doesn't work if the egress gateway is the local node. In that case it will send the packet to `cilium_vxlan` even though it should just leave through the native device. @jibi traced it as:
```
<- endpoint 1119 flow 0x96be6ada identity 36700->unknown state new ifindex 0 orig-ip 0.0.0.0: 10.11.1.194:54038 -> 192.168.33.13:1234 tcp SYN
-> overlay flow 0x96be6ada identity 36700->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.11.1.194:54038 -> 192.168.33.13:1234 tcp SYN
<- overlay flow 0x96be6ada identity unknown->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.11.1.194:54038 -> 192.168.33.13:1234 tcp SYN
-> stack flow 0x96be6ada identity unknown->unknown state new ifindex cilium_host orig-ip 0.0.0.0: 10.11.1.194:54038 -> 192.168.33.13:123  tcp SYN
-> network flow 0x96be6ada identity unknown->unknown state new ifindex 0 orig-ip 0.0.0.0: 192.168.33.12:54038 -> 192.168.33.13:1234 tcp SYN
```
We can see the packet is already SNATed with the wrong IP when it leaves the native device (`-> network`). The egress gateway IP was 192.168.33.100; the packet leaves with 192.168.33.12.

To fix this, we could set `tunnel_endpoint` to 0 from userspace when the `tunnel_endpoint` is the local node. Then, in the datapath, we can send such packets directly to the stack (assuming the stack won't SNAT it):
```c
if (!info)
	goto skip_egress_gateway;
if (!info->tunnel_endpoint)
	goto pass_to_stack;
```

(The egress gateway will refuse dstIP inside the cluster, right?)

anfernee replied:
Is the reason why the packet gets SNATed with the wrong IP by iptables? If I ran it with `--enable-bpf-masquerade` it doesn't have the issue. 

Actually for we are thinking of treating `egress_ip` as `tunnel_endpoint`, so we don't need to know which node has the `egress_ip`. For example, we can do this on the nodes: 
```
cilium bpf egress add 10.11.1.194 192.168.33.13/32 192.168.33.100 192.168.33.100
```

pchaigno replied:
> Is the reason why the packet gets SNATed with the wrong IP by iptables? If I ran it with --enable-bpf-masquerade it doesn't have the issue.

We had not enabled BPF masquerading, so that's probably it. We could likely add a clear dependency on BPF masquerading for the first version and fatal if users tries to enable egress gateway without it.

anfernee replied:
sounds like a plan!

anfernee replied:
Hey Paul, I found this doesn't have to depend on `--enable-bpf-masquerade`. I found that adding the following rule will fix the issue. I don't see any harm to disable conntrack for cilium_host's input path. What do you think. 

```
-A CILIUM_PRE_raw -i cilium_host -j NOTRACK
```

pchaigno replied:
Wouldn't that cause us to skip any reverse SNAT for traffic coming from the tunnel and going to the hostns? It feels a bit hackish and we should probably fix the packet path instead. We currently go `lxc -> cilium_vxlan -> cilium_host -> eth0` when we should just have `lxc -> eth0`.

Maybe we can require `enable-bpf-masquerade` for the first version and improve after the feature freeze?

anfernee replied:
Confirmed that it is the case. This is probably the best approach now. I've applied the change already.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1015,6 +1016,16 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 #endif
 
+#if defined(ENABLE_EGRESS_GATEWAY)
+	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) != MARK_MAGIC_SNAT_DONE) {
```
Could you add a comment to explain this check? (It's to skip packets that were already SNATed at the bpf_lxc device if I understand correctly.)

anfernee replied:
Exactly, it's to skip the revnat'ed packet from bpf_lxc.

pchaigno commented on bpf/Makefile:
```diff
@@ -234,7 +235,7 @@ LXC_OPTIONS = \
 ifneq ("$(KERNEL)","49")
 MAX_LXC_OPTIONS += -DENABLE_IPV4_FRAGMENTS -DENABLE_NODEPORT -DENABLE_SESSION_AFFINITY \
-	-DENABLE_EXTERNAL_IP -DENABLE_HOSTPORT -DENABLE_LOADBALANCER
+	-DENABLE_EXTERNAL_IP -DENABLE_HOSTPORT -DENABLE_LOADBALANCER -DENABLE_EGRESS_GATEWAY
```
We also need the same for `MAX_HOST_OPTIONS` and `HOST_OPTIONS` since there is code depending on `ENABLE_EGRESS_GATEWAY` in `bpf_host`.

pchaigno commented on bpf/lib/egress_gw.h:
```diff
@@ -0,0 +1,68 @@
+	struct ipv4_nat_target target = {
+		.min_port = 32767,
+		.max_port = 43835,
+	};
```
Should this reuse the NodePort parameters?
https://github.com/cilium/cilium/blob/b3f03879f460011bd9f4b51d021933ab066393d3/bpf/lib/nodeport.h#L1504-L1506

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1015,6 +1016,16 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 
+#if defined(ENABLE_EGRESS_GATEWAY)
+	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) != MARK_MAGIC_SNAT_DONE) {
+		ret = egress_nat_fwd(ctx);
```
I think you will need a bit more code on the receiving end to handle the packet. We deployed this on GKE and traced the packets on the gateway node:
```
$ ks exec cilium-2mgb7 -- cilium bpf egress list
SRC IP & DST CIDR       EGRESS INFO
10.76.1.55 1.1.1.1/32   10.164.0.104 10.164.0.99
$ ks exec cilium-2mgb7 -- cilium monitor | grep "1.1.1.1:443"
level=info msg="Initializing dissection cache..." subsys=monitor
<- overlay flow 0x0 identity unknown->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.76.1.55:49950 -> 1.1.1.1:443 tcp SYN
-> stack flow 0x0 identity unknown->unknown state new ifindex cilium_host orig-ip 0.0.0.0: 10.76.1.55:49950 -> 1.1.1.1:443 tcp SYN
-> network flow 0x0 identity unknown->unknown state new ifindex 0 orig-ip 0.0.0.0: 10.164.0.104:49950 -> 1.1.1.1:443 tcp SYN
```
We can see the SYN arrive from the overlay as expected and make its way to the native device via the stack. However, it is SNATed by existing iptables rules (in `-t nat IP-MASQ`) before it reaches the native device with this `egress_nat_fwd` code.

The easiest way to fix this would probably be to `redirect()` the packets to the native device as they arrive via the tunnel. That'll require a FIB lookup to select the device however.

pchaigno replied:
Also, note that when the packet arrives from the tunneling device, it hits this line https://github.com/cilium/cilium/blob/1db771f25895ca8c830bd16336dc7f87df4d222a/bpf/bpf_overlay.c#L250
which redirects it to `cilium_host`. It should redirect to native devices instead.

anfernee replied:
> The easiest way to fix this would probably be to redirect() the packets to the native device as they arrive via the tunnel. That'll require a FIB lookup to select the device however. 

This is a good suggestion. This should be general to all the packets, not just for egress, right?

anfernee replied:
I am not too worried about the iptables rules inserted by IP masq agent, since cilium will have it implemented in epbf as well. But I agree, we should skip kernel stack as much as possible.

pchaigno replied:
> This is a good suggestion. This should be general to all the packets, not just for egress, right?

Yeah, I think so, at least to avoid asymmetric paths.

pchaigno commented on bpf/lib/egress_gw.h:
```diff
@@ -0,0 +1,68 @@
+
+	target.addr = info->egress_ip;
+
+	return snat_v4_process(ctx, NAT_DIR_EGRESS, &target, from_endpoint);
```
The egress gateway currently relies on (1) BPF NodePort and (2) tunneling mode. Could you add checks in `initKubeProxyReplacementOptions()` and `daemon/cmd/daemon_main.go` to disable the egress gateway feature with a warning if any of these two conditions isn't satisfied?

anfernee replied:
Will add check for these 2 options: `--kube-proxy-replacement` and `--enable-bpf-masquerade`

pchaigno replied:
Sounds good :+1:

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -687,7 +711,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 
 		ret = encap_and_redirect_lxc(ctx, tunnel_endpoint, encrypt_key,
-					     &key, SECLABEL, monitor);
+					    &key, SECLABEL, monitor);
```
I think this change was included by mistake.

pchaigno reviewed (changes requested):
Hey @anfernee :wave: 

We pair reviewed the pull request with @jibi and have a few comments below. We also tested it in the development VMs and in GKE with both a remote and a local gateway node.

As far as we can tell, the redirection through the tunnel works fine, but the SNAT doesn't always. The packets sometimes get SNATed by the stack before they reach the native device and its egress gateway code. There also seem to be a couple of weird packet paths (packets being redirected to `cilium_vxlan` or `cilium_host` when they should go to the native devices).

pchaigno commented on bpf/Makefile:
```diff
@@ -137,7 +137,9 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_EXTERNAL_IP:-DENABLE_DSR: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_IPSEC:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_EXTERNAL_IP:-DENABLE_DSR: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_HOSTPORT: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_EXTERNAL_IP:-DENABLE_HOSTPORT: \
```
This line is a duplicate of the previous one, no?

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1029,6 +1030,17 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+						      CTX_ACT_DROP,
+						      METRIC_EGRESS);
+	}
+#endif
```
This block of code is causing the program size issue when combined with other features used in this file (bandwidth manager, BPF NodePort, host firewall, etc.). Maybe we could tail call before calling `snat_v4_process` (it depends whether we need to "come back" to this function to execute `lb_handle_health` below)?

anfernee committed:
Move egress handler in tailcall

This avoids the bpf_host program grows too large which potentially could
fail the verifier.

This change also refactors nodeport_nat_fwd to make it handle all nat
forwarding cases.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1029,6 +1049,12 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 #endif
 
+#if defined(ENABLE_EGRESS_GATEWAY)
+	ep_tail_call(ctx, CILIUM_CALL_IPV4_HANDLE_EGRESS_GW);
```
I don't think that will work as we still want to execute the code below (BPF NodePort) if the packet doesn't match a connection requiring egress gateway SNATing. Since we don't come back from a tail call, that BPF NodePort code will be skipped.

Instead I was proposing we move only `snat_v4_process` to a tail call, because I think at that point we know the packet needs to be SNATed.

anfernee replied:
oh right, I should've add the if check here as well. how about just move `egress_nat_fwd` into tailcall? `snat_v4_process` is used in several places. I feel not all places are good for tailcalls (like in `tail_nodeport_nat_ipv4`)?

pchaigno replied:
I meant we could do a tail call from `handle_egress_nat_ipv4` to `tail_egress_nat_fwd` and the tail called program would just call `egress_nat_fwd`. We don't need to modify other places calling `egress_nat_fwd`. The `lookup_ip4_egress_endpoint` call needs to be before the tail call or we will skip BPF NodePort.

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -595,6 +595,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	} else if option.Config.EnableIPMasqAgent {
 		log.Fatalf("BPF ip-masq-agent requires --%s=\"true\" and --%s=\"true\"", option.Masquerade, option.EnableBPFMasquerade)
+	} else if option.Config.EnableEgressGateway {
+		log.Fatalf("Egress Gateway requires --%s=\"true\" and --%s=\"true\"", option.Masquerade, option.EnableBPFMasquerade)
```
I think we also need to check that tunneling is enabled. It's easy to forget about that and lose 10s of minutes trying to find what's broken as I just did :sweat_smile:

anfernee replied:
It also works when tunnel is disabled. In tunnel disabled mode, it also creates a tunnel interface only for forwarding traffic. https://github.com/cilium/cilium/pull/14723/files#diff-3bd10cddebdc5669fc7a472e02c78702a9bc59fee30d502d1cc87a84a58439d8R309-R314

pchaigno replied:
It didn't work the first time I tried, but maybe that was because I tried to use an egress IP different from the Internal IP :shrug: 

Seems to work now.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -679,6 +679,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
+		 * path to be consistent. In future, it can be optimized by directly
+		 * direct to external interface.
+		 */
+		ret = encap_and_redirect_lxc(ctx, info->tunnel_endpoint, encrypt_key,
```
When the gateway node is the local node, this code causes us to send the packet to cilium_vxlan to then come back to the stack and finally leave through the native device. This is of course unnecessary and likely to cause issues later on.

@anfernee Is this something you want to fix in this PR? If not, we should track it somewhere. I think it's fine if we fix after the feature freeze.

anfernee replied:
Sure. We can treat it as an optimization. I've created #15426. The problem here is it's hard to tell it's tunneling to itself from here, because we are not using the `node_ip` as `tunnel_endpoint` here. It's the `egress_ip`. It doesn't know it owns the egress IP.

pchaigno reviewed (commented):
I ran tests again on GKE with both local and remote gateway nodes. It works fine, but only because I'm using the Internal IP of the gateway node as the egress IP (i.e., not adding a new IP). The issue is that the following `IP-MASQ` rule masquerades the packet before it even reaches the native device. So the egress gateway never gets a chance to SNAT the packet.
```
[1123:67360] -A IP-MASQ -m comment --comment "ip-masq: outbound traffic is subject to MASQUERADE (must be last in chain)" -j MASQUERADE
```

Did you test on GKE? Is this expected?

Do we want to fix this in the PR or should we take a look after the feature freeze?

pchaigno reviewed (approved):
It looks ready to me. I tested on GKE only, with an egress IP equal to the tunnel endpoint.

I think we're only missing the tests in the development VM with a proper egress IP (cc @jibi) and we should be good to merge.
